src/fields/field_add_potential.hh

```diff
@@ -22,6 +22,87 @@
 #include <memory>
 
 #include "fields/field.hh"
+#include "fields/field_algo_base.hh"
+#include "fields/field_model.hh"
+#include "fields/field_coords.hh"
+
+
+/*******************************************************************************
+ * Functors of FieldModels
+ */
+using Sclr = double;
+using Vect = arma::vec3;
+
+// Functor computing piezo_head_p0
+struct fn_add_potential {
+	inline Sclr operator() (Vect gravity, Vect coords, Sclr pressure) {
+        return arma::dot(gravity, coords) + pressure;
+    }
+};
+
+
+/**
+ * Factory class (descendant of @p Field<...>::FactoryBase) that is necessary
+ * for setting pressure values are piezometric head values.
+ */
+template <int spacedim, class Value>  // <3, FieldValue<3>::Scalar>
+class AddPotentialFactory : public Field<spacedim, Value>::FactoryBase {
+public:
+    /// Constructor.
+    AddPotentialFactory( Field<3, FieldValue<3>::VectorFixed > &gravity, FieldCoords &coords, Field<3, FieldValue<3>::Scalar> &inner_field)
+    : gravity_(gravity),
+	  coords_(coords),
+	  inner_field_(inner_field),
+      field_name_(inner_field.input_name())
+    {}
+
+    typename Field<spacedim,Value>::FieldBasePtr create_field(Input::Record rec, const FieldCommon &) override {
+        Input::AbstractRecord field_a_rec;
+        if (rec.opt_val(field_name_, field_a_rec)) {
+
+            FieldAlgoBaseInitData init_data(field_name_, Value::NRows_, UnitSI::dimensionless());
+            auto inner_field_ptr = FieldAlgorithmBase<spacedim, Value>::function_factory( field_a_rec, init_data );
+
+            // get domain specification
```

Comment:
Deduplicate processing of the region and rid, use correct time.
- Attribute field_input_name_ in FactoryBase, pass in constructor.
- Drop the parameter input_name of method is_active_field_descriptor.
- change parameters of FieldBase::create_field(descriptor_rec, time, const RegionSet &domain, ?? FieldCommon)
- check if passing field is necessary
- deduplicate create_field and Update history using these new parameters

Reply:
What exactly is the problem? The shared->input_name_ is set in FieldCommon::name() which is set during creation process of FieldCommon. Give more details.

Reply:
Yes, shared->input_name_ is set in FieldCommon::name(), but FactoryBase is create in default constructor of Field<> (see field.impl.hh, line 48) and 'input_name' parameter is not to disposition in this moment. FieldCommon::name() method is called later.